designdocs/CAPACITY.md

```diff
@@ -0,0 +1,148 @@
+pub type CapacityOf<T: Config> = StorageMap<_, Blake2_128Concat, MessageSenderId, Details<T::Balance>>;
+
+pub struct CapacityDetails {
+	pub used: Balance,
+	pub total_available: Balance,
+
+  // these to be discussed below
+  pub last_replenished_epoch: u32,
+  pub initial_epoch: u32
+}
+```
+
+And we can store the a token account staking to MSA by mapping.
+
+```rust
+#[pallet::storage]
+pub type Staking<T: Config> = StorageMap<_, Blake2_128Concat, T::AccountId, StakingDetails<T::Balance>>;
+
+pub struct StakingDetails<Balance> {
+	pub msa_id: MessageSenderId,
+	pub total_reserved: Balance,
+}
+```
+
+To give the limitations of what you can do with Capacity you are limited to staking and un-staking Capacity for token. Note that these interface can change as we introduce stash and controller accounts using Proxy pallet.
+
+```rust
+pub fn stake(origin: OriginFor<T>, account: MessageSenderId, amount: BalanceOf<T>)
+
+pub fn unstake(origin: OriginFor<T>, account: MessageSenderId, amount: BalanceOf<T>)
+```
+
+Events for triggers for staking include
+
+```rust
+pub enum Event<T: Config> {
+  Staked { account: T::AccountId, msa_id: MessageSenderId, amount: BalanceOf<T> },
+
+  UnStake{ account: T::AccountId, msa_id: MessageSenderId, amount: BalanceOf<T> },
+}
+```
+
+Errors when staking
+
+```rust
+pub enum Error<T> {
+	AlreadyStaked,
+	InsufficientBalance,		
+	InvalidMsa,
+}
+```
+
+**Capacity is replenishable**
+
+Replenishable means that after a certain fixed period, called an epoch, all capacity is replenished. The epoch period is composed of blocks.  Moreover, one unit of Capacity is `10^12` and matches with substrate weight measurement where `10^12` is 1 second of execution time.
+
+Since we want the ability to replenish Capacity the following interface is implement to replenish capacity.
+
+```rust
+
+traits Replenishable {
+   type Balance: Balance;
+   
+	 fn replenish_by_account(msa_id: MessageSenderId, amount: Balance) -> Result<Balance, Eroor> {};
+   fn replenish_all_for_account(msa_id: MessageSenderId) -> Result<Balance, Error>;
+   fn can_replenish(msa_id: MessageSenderId) -> bool;
+}
+```
+
+**How capacity replenishes?**
+
+As mentioned above, epoch periods are composed of blocks. The next epoch period calculated is based on the current epoch fullness, and expands and contracts depending on a threshold. This threshold is configurable and can be called `config::epochThreshold` . Additionally, we can also configure a multiplier that we increase or decrease the next epoch based on the fullness of the current epoch. This is defined as `config::epochUtilizationMultiplier`.
+
+![Untitled](https://user-images.githubusercontent.com/3433442/171747512-651112be-bbdc-4197-96ef-0cd208f702db.png)
+
+The above illustrates two epochs where the second one contracts because network congestion has decreased. As a result of epoch decreasing, capacity is replenish faster. Note that initially epoch periods will be fixed in sized for a period of time.
```

Comment:
> Note that initially epoch periods will be fixed in sized for a period of time.

please explain more on this

Reply:
betanet: 10-100 blocks
testnet: 5 blocks
mainet: 10-100 blocks